File: python/examples/molalign_rmsd.py
"""Read-only RMSD measurement and explicit molecular alignment."""

import numpy as np

import cosmolkit as ck

reference_coordinates = np.array(
    [[0.0, 0.0, 0.0], [1.0, 0.0, 0.0], [0.0, 2.0, 0.0]]
)
probe_coordinates = reference_coordinates + np.array([3.0, -2.0, 1.0])

reference = ck.Molecule.from_smiles("CCC").with_only_3d_conformer(reference_coordinates)
probe = ck.Molecule.from_smiles("CCC").with_only_3d_conformer(probe_coordinates)
params = ck.AlignmentParameters(
    atom_map=[ck.AlignmentAtomMap(index, index) for index in range(3)]
)

transform_result = probe.alignment_transform_to(reference, params)
aligned_probe, applied_result = probe.with_alignment_to(reference, params)

print("read-only RMSD:", transform_result.rmsd())
print("applied RMSD:", applied_result.rmsd())
print("source unchanged:", np.array_equal(probe.coordinates_3d(), probe_coordinates))
print("aligned coordinates:\n", aligned_probe.coordinates_3d())

File: crates/cosmolkit/examples/molalign_rmsd.rs
use cosmolkit::{AlignmentAtomMap, AlignmentParameters, Molecule};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let reference_coordinates = vec![[0.0, 0.0, 0.0], [1.0, 0.0, 0.0], [0.0, 2.0, 0.0]];
    let probe_coordinates = reference_coordinates
        .iter()
        .map(|point| [point[0] + 3.0, point[1] - 2.0, point[2] + 1.0])
        .collect();

    let reference = Molecule::from_smiles("CCC")?.with_only_3d_conformer(reference_coordinates, true)?;
    let probe = Molecule::from_smiles("CCC")?.with_only_3d_conformer(probe_coordinates, true)?;
    let params = AlignmentParameters {
        atom_map: Some(
            (0..3)
                .map(|index| AlignmentAtomMap {
                    probe_atom: index,
                    reference_atom: index,
                })
                .collect(),
        ),
        ..AlignmentParameters::default()
    };

    let measurement = probe.alignment_transform_to(&reference, &params)?;
    let (aligned, applied) = probe.with_alignment_to(&reference, &params)?;

    println!("read-only RMSD: {}", measurement.rmsd);
    println!("applied RMSD: {}", applied.rmsd);
    println!(
        "source first coordinate: {:?}",
        probe.conformers_3d()[0].coordinates()[0]
    );
    println!(
        "aligned first coordinate: {:?}",
        aligned.conformers_3d()[0].coordinates()[0]
    );
    Ok(())
}
